src/Microsoft.VisualStudio.LanguageServices.Razor/ProjectSystem/FallbackRazorProjectHost.cs

```diff
@@ -0,0 +1,143 @@
+    //
+    // This class is responsible for intializing the Razor ProjectSnapshotManager for cases where
+    // MSBuild does not provides configuration support (< 2.1).
+    [AppliesTo("DotNetCoreRazor & !DotNetCoreRazorConfiguration")]
+    [Export(ExportContractNames.Scopes.UnconfiguredProject, typeof(IProjectDynamicLoadComponent))]
+    internal class FallbackRazorProjectHost : RazorProjectHostBase
+    {
+        private const string MvcAssemblyName = "Microsoft.AspNetCore.Mvc.Razor";
+        private const string MvcAssemblyFileName = "Microsoft.AspNetCore.Mvc.Razor.dll";
+
+        private IDisposable _subscription;
+
+        [ImportingConstructor]
+        public FallbackRazorProjectHost(
+            IUnconfiguredProjectCommonServices commonServices,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace)
+            : base(commonServices, workspace)
+        {
+        }
+
+        // Internal for testing
+        internal FallbackRazorProjectHost(
+            IUnconfiguredProjectCommonServices commonServices,
+             Workspace workspace,
+             ProjectSnapshotManagerBase projectManager)
+            : base(commonServices, workspace, projectManager)
+        {
+        }
+
+        protected override async Task InitializeCoreAsync(CancellationToken cancellationToken)
+        {
+            await base.InitializeCoreAsync(cancellationToken).ConfigureAwait(false);
+
+            // Don't try to evaluate any properties here since the project is still loading and we require access
+            // to the UI thread to push our updates.
+            //
+            // Just subscribe and handle the notification later.
+            var receiver = new ActionBlock<IProjectVersionedValue<IProjectSubscriptionUpdate>>(OnProjectChanged);
+            _subscription = CommonServices.ActiveConfiguredProjectSubscription.JointRuleSource.SourceBlock.LinkTo(
+                receiver,
+                initialDataAsNew: true,
+                suppressVersionOnlyUpdates: true,
+                ruleNames: new string[] { ResolvedCompilationReference.SchemaName });
+        }
+
+        protected override async Task DisposeCoreAsync(bool initialized)
+        {
+            await base.DisposeCoreAsync(initialized).ConfigureAwait(false);
+
+            if (initialized)
+            {
+                _subscription.Dispose();
+            }
+        }
+
+        // Internal for testing
+        internal async Task OnProjectChanged(IProjectVersionedValue<IProjectSubscriptionUpdate> update)
+        {
+            await ExecuteWithLock(async () =>
+            {
+                if (IsDisposing || IsDisposed)
+                {
+                    return;
+                }
+
+                string mvcReferenceFullPath = null;
+                var references = update.Value.CurrentState[ResolvedCompilationReference.SchemaName].Items;
```

Comment:
Wondering if we should add an API that has this "is valid host project" logic. When it comes to VS4Mac we'll have to introduce these concepts all over again in a non-CPS way.

Not saying this is necessary for this PR but more food for thought on the final forms of these APIs.

Reply:
I'd be interested to know what you're thinking about. 

Right now I'm thinking about the project system and the data flow therein, so I'm not trying to work very hard to address the editor's needs. 

Here's what I'm thinking of:
- **every** open dotnet core razor document open in the editor should belong to a project
- if a dotnet core razor document needs to open and doesn't have a project to belong to, the snapshot manager will create an *ephemeral* project
- the snapshot manager can remove an *ephemeral* project when all documents are closed
- the snapshot manager can *reify* the *ephemeral* project by getting an update from the underlying project system

This means that the project system:
A. Needs no knowledge about editors and open documents
B. Only needs to reflect the underlying truth of the host's project system

This means that the editor:
A. Can always have a project to work with
B. Can get 'canned' answers about tag helpers and such when the project is *ephemeral*
C. Doesn't really care if the project is real or not

All of the heavy lifting is in the snapshot manager/document manager, of which we only have one implementation.

Thoughts?

Reply:
> All of the heavy lifting is in the snapshot manager/document manager, of which we only have one implementation.
> 
> Thoughts?

Love the separation you're bringing up. Depending on how much logic ends up sitting on the `HostProject` side of the world I might want to consider building a cross-platform pattern way of orchestrating `HostProject` => snapshot/document manager interaction. 

Right now I think we have a minimal set of things going on in HostProject land which is awesome; however, things like falling back to older versions of MVC is something that every host project system will have to understand.